source/ice_api.c

```diff
@@ -0,0 +1,1676 @@
+    }
+
+    if( retStatus == ICE_RESULT_OK )
+    {
+        *ppSendStunMessageBuffer = &pIceAgent->stunMessageBuffer[ 0 ];
+    }
+
+    return retStatus;
+}
+/*------------------------------------------------------------------------------------------------------------------*/
+
+/* Ice_DeserializeStunPacket - This API deserializes a received STUN packet . */
+
+IceResult_t Ice_DeserializeStunPacket( StunContext_t * pStunCxt,
+                                       StunHeader_t * pStunHeader,
+                                       StunAttribute_t * pStunAttribute,
+                                       IceStunDeserializedPacketInfo_t * pDeserializedPacketInfo )
+{
+    IceResult_t retStatus = ICE_RESULT_OK;
+    StunResult_t stunRetStatus = STUN_RESULT_OK;
+    uint8_t * pErrorPhase;
+    uint16_t errorPhaseLength;
+
+    if( ( pStunCxt == NULL ) ||
+        ( pStunHeader == NULL ) ||
+        ( pStunAttribute == NULL ) ||
+        ( pDeserializedPacketInfo == NULL ) )
+    {
+        retStatus = ICE_RESULT_BAD_PARAM;
+    }
+
+    /* Initialization */
+    pDeserializedPacketInfo->useCandidateFlag = 0;
+    pDeserializedPacketInfo->errorCode = 0;
+    pDeserializedPacketInfo->priority = 0;
+
+    while( stunRetStatus == STUN_RESULT_OK )
+    {
+        stunRetStatus = StunDeserializer_GetNextAttribute( pStunCxt,
+                                                           pStunAttribute );
+
+        if( stunRetStatus == STUN_RESULT_OK )
+        {
+            switch( pStunAttribute->attributeType )
+            {
+                case STUN_ATTRIBUTE_TYPE_ERROR_CODE:
+                    stunRetStatus = StunDeserializer_ParseAttributeErrorCode( pStunAttribute,
+                                                                              &pDeserializedPacketInfo->errorCode,
+                                                                              &pErrorPhase,
+                                                                              &errorPhaseLength );
+                    break;
+
+                case STUN_ATTRIBUTE_TYPE_XOR_MAPPED_ADDRESS:
+                    stunRetStatus = StunDeserializer_ParseAttributeAddress( pStunCxt,
+                                                                            pStunAttribute,
+                                                                            &pDeserializedPacketInfo->stunAttributeAddress );
+                    break;
+
+                case STUN_ATTRIBUTE_TYPE_USE_CANDIDATE:
+                    pDeserializedPacketInfo->useCandidateFlag = 1;
+                    break;
+
+                case STUN_ATTRIBUTE_TYPE_PRIORITY:
+                    stunRetStatus = StunDeserializer_ParseAttributePriority( pStunCxt,
+                                                                             pStunAttribute,
+                                                                             &pDeserializedPacketInfo->priority );
+                    break;
+
```

Comment:
Should ICE component verify the integrity? Can we say the ICE component's functionality is aligned with current implementation of WebRTC SDK?

Reply:
updated.